### ISSUE TITLE
Implementing InnerJoin with HashMap

### ISSUE THREAD

**Issue by ksranga** [jOOQ/jOOL#134]
Hi - do you see a performance benefit in implementing the InnerJoin by doing a lookup against HashMap after converting one of the List to a Map using the join column as the key?

**Comment by lukaseder** [comment]
Indeed, implementing a hash join rather than a nested loop might speed things up in certain situations. Although, care must be taken in the event of duplicates. It's perfectly fine for an inner join to operate on duplicates, e.g. the following test must pass:

```java
        assertEquals(asList(
            tuple(1, 1),
            tuple(1, 2),
            tuple(1, 1),
            tuple(1, 2)),
            Seq.of(1, 1).innerJoin(Seq.of(1, 2), (a, b) -> true).toList());
```

Would you be interested in implementing this improvement?

**Comment by ksranga** [comment]
Yes - I can help out. To handle duplicates, the Map we have to build would have a List as its value

Map<TupleN, List<T>>

TupleN are the columns in the join condition.

**Comment by lukaseder** [comment]
Indeed, this would work. I suspect that we should implement this optimisation only after a given threshold, e.g. when the argument list size exceeds 8 or so, as the map creation and hashcode / equals overhead might exceed the benefit of using a map

**Comment by ksranga** [comment]
I am just wrapped up this feature. How do I go about getting access so I can create a pull request?

**Comment by lukaseder** [comment]
Just create a pull request... In order to do so, commit your changes to your own fork, and click on the pull request button on GitHub

**Comment by lukaseder** [comment]
I'm closing this as "won't fix". While with sufficent knowledge about the nature of the joined sets that participate in a join operation (as in SQL, when indexes and constraints are known), a nested loop might just be good enough in a more general use-case.

The suggestions exposed in this discussion are not doable, as they

1. Change the semantics of a relational join operation in our API
2. Change the API itself
